designs/nodeclaim-limits.md

```diff
@@ -0,0 +1,54 @@
+# NodeClaim Limits
+
+**NodeClaim**: Karpenter’s representation of a Node request. A NodeClaim declaratively describes how the node should look, including instance type requirements that constrain the options that the Cloud Provider can launch with. For the AWS Cloud Provider, the instance requirements of the NodeClaim translate to the requirements of the CreateFleet request. More details can be found [here](https://github.com/aws/karpenter-provider-aws/blob/main/designs/node-ownership.md).
+
+## Motivation
+
+The Karpenter NodePool Spec allows for specifying [provisioning limits](https://karpenter.sh/docs/concepts/nodepools/#speclimits) which can be used to limit based on any resource type provided by the cloudprovider.  If a limit is exceeded, node provisioning is prevented until more resources become available (e.g. node terminations).
+
+This works for users who are purely concerned about limiting resource usage within a NodePool, but doesn't work for specifying a global limit across all NodePools or for limiting a higher level representation of usage such as the number of nodes in a cluster (without also reducing flexibility of instance types).
+
+## Use-Cases
+1. I'm using a shared network with limited IP pools and need to ensure Karpenter's usage doesn't consume too many IPs. Each Node consumes a minimum number of IPs but low-level resource usage for things like CPU/Memory differs between instance types.
+2. I have a special NodePool (for example a set of high performance nodes or AWS Reserved Instances) with a finite amount of compute available. I need a way to prevent Karpenter from launching too many nodes of this type. The only to do this is by calculating against an avialable resource like number of CPU's and adding this as a limit. This doesn't target the actual limiter which is number of instances rather than resource consumption.
+3. My CNI can't support more than 400 nodes but I'm not concerned about resource usage. I have no way to restrict Karpenter to this limit without severly limiting instance type options.
+
+
+Currently, Karpenter users face a limitation in controlling the maximum number of nodes provisioned by Karpenter. This constraint hinders operations in environments where the CNI in use can't scale beyond a certain number of nodes or in scenarios involving shared networks with a finite IP address allocation. Some additional use cases and discussion on this limitation can be seen in the following issues: (#732, aws/karpenter-provider-aws#4462).
+
+## Solutions
+
+### [Recommended] Solution 1: New controller flag `max-node-claims` and per-NodePool limits in the `spec.limits` section of NodePool template
+
+1. A new controller startup flag `max-node-claims` which would allow for setting a global cap on the number of nodes provisioned via this Karpenter controller.
+2. Allow for setting per-NodePool limits on number of NodeClaims. 
```

Comment:
Why would we need both? It seems sufficient that users could set the nodepool level. Global limits are a different and probably are better encapsulated with a different RFC/feature request as shown here: https://github.com/kubernetes-sigs/karpenter/issues/745

Reply:
Nodepool level isn't sufficent as there are hard limits at the cluster level  in some cases, but I agree it is better encapsulated with the proper cluster level limits feature request. 

These limits should be part of that ratehr than its own flag.

Reply:
Hi, yes it's desirable to have a global limit for users that have constraints at the cluster level.
Would it be possible to keep the scope of this proposal to cover a mechanism for limiting number of nodes?

Reply:
I would say the real design choice here would be deciding if we should add a flag for "max-node-claims", or should we have global limits encapsulated in the design shared by njtran. 

Otherwise, later we might end up adding the node limits to this global mechanism and then later have to change it.

It is better to get the configuration right from the start.

Reply:
Agreed on getting it right the first time. Is there a design that's been discussed for global limits or do you just mean the feature request in https://github.com/kubernetes-sigs/karpenter/issues/745?

Reply:
We avoid global configuration in Karpenter as a general rule. Is a policy enforcement engine + NodePool limits a better approach?